utils/proxy/proxyconfig.go

```diff
@@ -0,0 +1,172 @@
+	defer pc.mu.Unlock()
+
+	var proxy *url.URL
+	if req.URL.Scheme == "https" {
+		proxy = pc.https
+	}
+	if proxy == nil {
+		proxy = pc.http
+	}
+	if proxy == nil {
+		return nil, nil
+	}
+	if !pc.useProxy(canonicalAddr(req.URL)) {
+		return nil, nil
+	}
+	return proxy, nil
+}
+
+// useProxy reports whether requests to addr should use a proxy,
+// according to the NO_PROXY or no_proxy environment variable.
+// addr is always a canonicalAddr with a host and port.
+func (pc *ProxyConfig) useProxy(addr string) bool {
+	if len(addr) == 0 {
+		return true
+	}
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return false
+	}
+	if host == "localhost" {
+		return false
+	}
+	if ip := net.ParseIP(host); ip != nil {
+		if ip.IsLoopback() {
+			return false
+		}
+	}
+
+	if pc.noProxy == "*" {
+		return false
+	}
+
+	addr = strings.ToLower(strings.TrimSpace(addr))
+	if hasPort(addr) {
+		addr = addr[:strings.LastIndex(addr, ":")]
+	}
+
+	for _, p := range strings.Split(pc.noProxy, ",") {
+		p = strings.ToLower(strings.TrimSpace(p))
+		if len(p) == 0 {
+			continue
+		}
+		if hasPort(p) {
+			p = p[:strings.LastIndex(p, ":")]
+		}
+		if addr == p {
+			return false
+		}
+		if p[0] == '.' && (strings.HasSuffix(addr, p) || addr == p[1:]) {
+			// no_proxy ".foo.com" matches "bar.foo.com" or "foo.com"
+			return false
+		}
+		if p[0] != '.' && strings.HasSuffix(addr, p) && addr[len(addr)-len(p)-1] == '.' {
+			// no_proxy "foo.com" matches "bar.foo.com"
+			return false
+		}
+	}
+	return true
+}
+
+// InstallInDefaultTransport sets the proxy resolution used by the
+// default HTTP transport to use the proxy details stored in this
+// ProxyConfig. Requests made without an explicit transport will
+// respect these proxy settings.
+func (pc *ProxyConfig) InstallInDefaultTransport() error {
+	transport, ok := http.DefaultTransport.(*http.Transport)
+	if !ok {
+		return errors.Errorf("http.DefaultTransport was %T instead of *http.Transport", http.DefaultTransport)
+	}
+	transport.Proxy = pc.GetProxy
+	return nil
+}
+
+var DefaultConfig = ProxyConfig{}
+
+func tolerantParse(value string) (*url.URL, error) {
+	if value == "" {
+		return nil, nil
+	}
+	proxyURL, err := url.Parse(value)
+	if err != nil || !strings.HasPrefix(proxyURL.Scheme, "http") {
+		// proxy was bogus. Try prepending "http://" to it and
+		// see if that parses correctly. If not, we fall
+		// through and complain about the original one.
+		if proxyURL, err := url.Parse("http://" + value); err == nil {
+			return proxyURL, nil
+		}
+	}
+	if err != nil {
+		return nil, errors.Errorf("invalid proxy address %q: %v", value, err)
+	}
+	return proxyURL, nil
+}
+
+// Internal utilities copied from net/http/transport.go
+var portMap = map[string]string{
+	"http":  "80",
+	"https": "443",
+}
+
+// canonicalAddr returns url.Host but always with a ":port" suffix
+func canonicalAddr(url *url.URL) string {
+	addr := url.Host
+	if !hasPort(addr) {
+		return addr + ":" + portMap[url.Scheme]
```

Comment:
This is technically wrong (in the stdlib too!). It generates invalid IPv6 address:port strings (we've screwed this up before). It looks like the code gets away with it because SplitHostPort allows the badly formatted values.

Use https://golang.org/pkg/net/#JoinHostPort instead

I see it's fixed upstream: https://go.googlesource.com/go/+/master/src/net/http/transport.go#1990